### PR TITLE
Fix: make dry run optional for patch task instance 

### LINKED ISSUE #34563: DryRun is not optional for patch task instance
## Summary

According to the [REST api docs](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/patch_mapped_task_instance).
you can patch a task instance state. When you hit the api without sending a
"dry_run" variable, you get a KeyError (This is from a server running version 2.5.3):

```
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/airflow/.local/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/airflow/.local/lib/python3.10/site-packages/connexion/decorators/decorator.py", line 68, in wrapper
    response = function(request)
  File "/home/airflow/.local/lib/python3.10/site-packages/connexion/decorators/uri_parsing.py", line 149, in wrapper
    response = function(request)
  File "/home/airflow/.local/lib/python3.10/site-packages/connexion/decorators/validation.py", line 196, in wrapper
    response = function(request)
  File "/home/airflow/.local/lib/python3.10/site-packages/connexion/decorators/validation.py", line 399, in wrapper
    return function(request)
  File "/home/airflow/.local/lib/python3.10/site-packages/connexion/decorators/response.py", line 112, in wrapper
    response = function(request)
  File "/home/airflow/.local/lib/python3.10/site-packages/connexion/decorators/parameter.py", line 120, in wrapper
    return function(**kwargs)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/api_connexion/security.py", line 51, in decorated
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/api_connexion/endpoints/task_instance_endpoint.py", line 594, in patch_task_instance
    if not data["dry_run"]:
KeyError: 'dry_run'
```

The API docs state that dry_run is not required and that it is defaulted to false.

This can be reproduced in `main` with the tests by commenting out line 1699 in
[test_task_instance_endpoint.py](https://github.com/apache/airflow/blob/5b0ce3db4d36e2a7f20a78903daf538bbde5e38a/tests/api_connexion/endpoints/test_task_instance_endpoint.py#L1695-L1709)

### PR DESCRIPTION
**Pull request by Calder-Ty**
Fixes #34563. This sets Dry run to be optional as the docs indicate it should be and makes the default False per the docs.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by Calder-Ty** [comment]
This is a reopened version of #34564 which was closed by mistake (by myself).

**Comment by Calder-Ty** [comment]
I didn't want this comment to be forgotten:

https://github.com/apache/airflow/pull/34564#pullrequestreview-1640562306

**Comment by f0restron07** [comment]
Hi @Calder-Ty,

Thank you for addressing this issue and for reopening the pull request. Making dry_run optional and setting the default to False aligns the behavior of the code with the documentation, which will certainly prevent confusion and errors for API users in the future.

It is crucial that the API's behavior is consistent with the documentation to ensure a smooth and reliable user experience, and this change is a step in the right direction.

Looking at the commits, it seems you have kept the changes concise and directly related to the issue at hand, which is appreciated. It's also good to see that the checks have mostly passed, indicating that this change doesn't introduce any apparent regressions or new issues.

I look forward to seeing this change merged after the required reviews, and I believe it will be a valuable improvement to Apache Airflow.

Thank you again for your contribution!

Best regards,
f0restron

**Comment by hussein-awala** [comment]
@pierrejeambrun in the first PR, I commented with:
> IMHO it's a problem in the doc, dry_run should be True by default.

WDYT? If you agree, should we fix the doc and the bug in the same PR?

**Comment by pierrejeambrun** [comment]
Yes, all other `dry_run` params default to True in other routes, the only 2 that do not in the doc are:
- https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/patch_task_instance
- https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/patch_mapped_task_instance

I think defaulting to true makes more sense, and is more conservative, unless explicitly stated by the user with a dry_run=False, we should run in dry run mode.

Also just for a consistency reason, dry_run should be True. Updating the doc and the schema will in one place will fix the both endpoints mentioned.

**Comment by Calder-Ty** [comment]
made changes to have docs and code aligned, as well as be consistent with the rest of the API (i.e Set `dry_run` to be true by default,

File: airflow/api_connexion/openapi/v1.yaml [hunk 1]
```diff
@@ -4302,7 +4302,7 @@ components:
             If set, don't actually run this operation. The response will contain the task instance
             planned to be affected, but won't be modified in any way.
           type: boolean
-          default: false
+          default: true
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml:4305]: This should be false

File: airflow/api_connexion/schemas/task_instance_schema.py [hunk 2]
```diff
@@ -180,7 +180,7 @@ def validate_form(self, data, **kwargs):
 class SetSingleTaskInstanceStateFormSchema(Schema):
     """Schema for handling the request of updating state of a single task instance."""
 
-    dry_run = fields.Boolean(dump_default=True)
+    dry_run = fields.Boolean(dump_default=True, load_default=True)
```
Comment (ephraimbuddy) [inline airflow/api_connexion/schemas/task_instance_schema.py]: Not needed.

**Review by ephraimbuddy — COMMENTED** [review]
We should only update here: https://github.com/apache/airflow/blob/d1b7bca6a36c146119fb5746019116c4d1e15275/airflow/api_connexion/endpoints/task_instance_endpoint.py#L636 to use `data.get(‘dry_run’, False)`

File: airflow/api_connexion/openapi/v1.yaml (hunk 1 above)
Comment (Calder-Ty) [reply airflow/api_connexion/openapi/v1.yaml:4305]: Per the conversation with @pierrejeambrun and @hussein-awala, I changed it to true

File: airflow/api_connexion/schemas/task_instance_schema.py (hunk 2 above)
Comment (Calder-Ty) [reply airflow/api_connexion/schemas/task_instance_schema.py]: Respectfully, I think this is a better fix than using `.get('dry_run', False)`. The Schema is a more natural place to define defaults for the request input. It also ensures that dry_run will _always_ have a value. The alternative does not help if in the future someone comes and tries to do a raw indexing access again. Anytime the `dry_run` parameter is accessed we (or another contributor who is unfamiliar with this conversation) would have to know that `dry_run` could possibly not exist as a key and the the default ought to be `False` (or `True`, depending on how that conversation is decided).

**Review by pierrejeambrun — APPROVED** [review]
LGTM, this is consistent with `dry_run` default value for other endpoints. (And the more conservative choice, I believe)

**Review by o-nikolas — APPROVED** [review]
Other than the static check failures (should be an easy fix to the generated ts) it lgtm.

I like the fix in the schema!

**Comment by o-nikolas** [comment]
Also, are we considering this change to the default a breaking change to the API or a bugfix?

CC @eladkal

**Comment by Calder-Ty** [comment]
> Also, are we considering this change to the default a breaking change to the API or a bugfix?
> 
> CC @eladkal

Technically it is, however, I think we can be fairly sure no one was depending on the functionality since it was broken.

**Comment by hussein-awala** [comment]
> Also, are we considering this change to the default a breaking change to the API or a bugfix?
> 
> CC @eladkal

Based on this issue #34563, we cannot use the endpoint without providing the dry_run value, so this PR technically adds a new default value for dry_run and not changing it from False to True.

File: airflow/api_connexion/schemas/task_instance_schema.py (hunk 2 above)
Comment (hussein-awala) [inline airflow/api_connexion/schemas/task_instance_schema.py]: Nit: if it works without `dump_default=True`, I prefer to remove it, it seems useless
```suggestion
    dry_run = fields.Boolean(load_default=True)
```

**Review by hussein-awala — APPROVED** [review]

**Comment by pierrejeambrun** [comment]
Yep, what Hussein said, not a breaking change

**Pull request merged**